### 协程 —— 协程的异步调用|挂起函数的组合

#### 1. 两个suspend一起调用会产生串行效果

```
fun main() = runBlocking {
    val elapsedTime = measureTimeMillis {
        val value1 = intValue1()
        val value2 = intValue2()
        println("$value1 + $value2 = ${value1 + value2}")
    }
    println("total time: $elapsedTime")
}

private suspend fun intValue1(): Int {
    delay(2000)
    return 15
}

private suspend fun intValue2() :Int {
    delay(3000)
    return 20
}
```

输出
```
15 + 20 = 35
total time: 5015
```

- intValue1() delay 2s，intValue2() delay 3s，因此最终elapsedTime > 5s。这说明两个函数是串行执行的，先执行完intValue1()，再执行intValue2()

**********************************

#### 2. 使用async与await实现并发

- [`async`](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/async.html)就像是launch一样，会开启一个单独的协程，可以与其他协程并发工作

- 区别：
    - launch会返回一个Job，但是Job并不会持有任何结果值；
    - async会返回一个Deferred，这是一个轻量级的非阻塞的future，它代表一个promise，可以在稍后提供一个结果值

- 可以通过在一个deferred值上调用[`await()`](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-deferred/await.html)方法来获取最终结果值，Deferred也是一个Job，因此可以在需要时对其进行取消

将1中代码用async和await做出更改
```
fun main() = runBlocking {
    val elapsedTime = measureTimeMillis {
        val value1 = async { intValue1() }
        val value2 = async { intValue2() }

        val result1 = value1.await()
        val result2 = value2.await()

        println("$result1 + $result2 = ${result1 + result2}")
    }

    println("total time: $elapsedTime")
}

private suspend fun intValue1(): Int {
    delay(2000)
    return 15
}

private suspend fun intValue2() :Int {
    delay(3000)
    return 20
}
```

输出
```
15 + 20 = 35
total time: 3015
```
- 利用async执行intValue1()和intValue2()，获取到两个Deferred对象，实现并发效果（elapsedTime > 3s)

**********************************

#### 3. async的延迟执行

- 通过将async方法的start参数设置为coroutineStart.LAZY来实现协程的延迟执行

- 此种情况下，协程会在两种情况下执行：调用Deferred的await方法，或是调用Job的start方法(Deferred是Job子类）

调用Job的start方法：

```
fun main() = runBlocking {
    val elapsedTime = measureTimeMillis {
        val value1 = async(start = CoroutineStart.LAZY) { intValue1() }       // (1)
        val value2 = async(start = CoroutineStart.LAZY) { intValue2() }

        println("hello world")

        Thread.sleep(6000)

        value1.start()                                                        // (2)
        value2.start()                                                        // (3)

        val result1 = value1.await()                                          // (4)
        val result2 = value2.await()                                          // (5)

        println("$result1 + $result2 = ${result1 + result2}")
    }

    println("total time: $elapsedTime")
}

private suspend fun intValue1(): Int {
    delay(2000)
    return 15
}

private suspend fun intValue2() :Int {
    delay(3000)
    return 20
}
```

结果：
```
hello world
15 + 20 = 35
total time: 9023
```

- (1) 处start赋值为LAZY，因此代码执行到此处时，intValue1()不会执行

- (2)、(3)处才开始执行intValue1()和intValue2()

- (4)、(5)处并发执行，使用时长为最长的函数时间，3s，因此elapsedTime > 6s + 3s = 9s

- 如果将(4)和(5)处代码互换位置呢，执行结果如何？

调用Deferred的await方法

```
fun main() = runBlocking {
    val elapsedTime = measureTimeMillis {
       val value1 = async(start = CoroutineStart.LAZY) { intValue1() }       // (1)
        val value2 = async(start = CoroutineStart.LAZY) { intValue2() }

        println("hello world")

        Thread.sleep(6000)

//        value1.start()                                                        // (2)
//        value2.start()                                                        // (3)

        val result1 = value1.await()                                          // (4)
        val result2 = value2.await()                                          // (5)

        println("$result1 + $result2 = ${result1 + result2}")
    }

    println("total time: $elapsedTime")
}

private suspend fun intValue1(): Int {
    delay(2000)
    return 15
}

private suspend fun intValue2() :Int {
    delay(3000)
    return 20
}
```

结果：
```
hello world
15 + 20 = 35
total time: 11020
```
- 注释掉了(2)和(3)，执行结果不同

- 代码执行到(1)时，intValue1()并未执行

- 由于(2)和(3)被注释掉，代码执行到(4)处等待2s（此时intValue2()并未执行），直到2s后，value1获取到了结果，代码(5)才开始执行

- 因此elapsedTime > 6s + 2s + 3s = 11s

- 换句话说，注释掉start()代码，此程序变为串行执行（因为此时async(start = CoroutineStart.LAZY)，程序的执行权交给了程序员控制）

**********************************

#### 4。 异步风格函数
将suspend函数封装在GlobalScope.async{}中，实现异步风格函数

```
fun main() {
    val elapsedTime = measureTimeMillis {
        val value1 = intValue1Async()                       // (1)
        val value2 = intValue2Async()
                                                            // (2)
        runBlocking {
            println("the answer is: ${value1.await()} + ${value2.await()}")
        }
    }

    println("total time: $elapsedTime")
}


private suspend fun intValue1(): Int {
    delay(2000)
    return 15
}

private suspend fun intValue2() :Int {
    delay(3000)
    return 20
}

fun intValue1Async() = GlobalScope.async {
    intValue1()
}

fun intValue2Async() = GlobalScope.async {
    intValue2()
}
```
结果
```
the answer is: 15 + 20
total time: 30732
```
- （1）处调用GlobalScope.async {}，其默认start=CoroutineStart.DEFAULT，立即执行。因此最终elapsedTime > 3s

-  此方法google不推荐，因为如果代码在(2)处有自定义逻辑，而且抛出异常，此时无法停止正在运行的协程。可以像下面例子一样，将代码放入coroutineScope内：

        fun main() = runBlocking {
            val elapsedTime = measureTimeMillis {
                println("answer is: ${intSum()}")
            }

            println("total sum: $elapsedTime")
        }

        private suspend fun intSum() : Int = coroutineScope {
            val value1 = async { intValue1() }
            val value2 = async { intValue2() }

            value1.await() + value2.await()
        }

        private suspend fun intValue1(): Int {
            delay(2000)
            return 15
        }

        private suspend fun intValue2() :Int {
            delay(3000)
            return 20
        }


**********************************

#### 5。 父子协程的异常与取消问题

- 当协程之间存在父子关系时，不管是取消还是异常，都会沿着协程的层次体系向上进行传播

```
fun main() = runBlocking<Unit> {                                    (1)
    try {
        failureComputation()
    } finally {
        println("computation failed")
    }
}

private suspend fun failureComputation() : Int = coroutineScope {
    val value1 = async<Int> {                                       (2)
        try {
            delay(9000000)
            50
        } finally {
            println("value1 was cancelled")
        }
    }

    val value2 = async<Int> {                                       (3)
        Thread.sleep(2000)
        println("value2 throws an exception")

        throw Exception()
    }

    value1.await() + value2.await()
}
```

输出
```
value2 throws an exception
value1 was cancelled
computation failed
Exception in thread "main" java.lang.Exception
```

- (1)处产生了一个协程，(2)和(3)处分别是协程(1)内部创建的两个新协程

- 当协程(3)抛出了一个异常，执行路径由(3)->(1)，因此输出"exception"和"failed"

- 协程(2)相当于被取消，因此也会执行finally代码块





